falcon/util/uri.py

```diff
@@ -292,9 +292,9 @@ def parse_query_string(query_string, keep_blank_qs_values=False):
             # a list, or append the next value to the list.
             old_value = params[k]
             if isinstance(old_value, list):
-                old_value.append(v)
+                old_value.append(decode(v))
             else:
-                params[k] = [old_value, v]
+                params[k] = [old_value, decode(v)]
 
         else:
             if ',' in v:
```

Comment:
At this point, I think we're just missing the decoding of the name (`k`), and improving the documentation per (3.b.ii).

@steffgrez Would you mind adding a line to decode `k` and a test for the same? Regarding (3.b.ii), we could update the note in the docstring: For example:

```
     Note:
         In addition to the standard HTML form-based method for specifying
         lists by repeating a given param multiple times, Falcon supports
         a more compact format in which the param may be given a single time
         but set to a ``list`` of comma-separated elements (e.g., 'foo=a,b,c').

         When using this format, if any values need to be uri-encoded, the 
         client should take care not to encode the commas used as 
         delimiters. Otherwise, Falcon will not treat the value as a list.
 
         The two different ways of specifying lists may not be mixed in
         a single query string for the same parameter.
```